File: nim/acronym/acronym.nim
import strutils

proc abbreviate*(input: string): string =
  var state = 'A'

  for ch in input:
    if state == 'A':
      # state A: seeking the next letter to add to the acronym
      if ch.isAlphaAscii:
        result.add(ch)
        state = 'N'

    else:
      # state N: seeking the end of this sequence of
      # letters and apostrophes, so we can begin
      # seeking the next letter.
      if (ch.isAlphaAscii or ch == '\'').not:
        state = 'A'

  result.toUpperAscii

File: dart/acronym/lib/acronym.dart
/* A word starts with a letter and can contain word chars and apostrophes.
 * I rely on regex "leftmost longest" matching to find the first letter of each word.
 */
final wordRegex = RegExp(r"(\p{Alpha})[\w']*", unicode: true);

class Acronym {
  String abbreviate(String input) =>
      wordRegex.allMatches(input).map((m) => m.group(1)).join('').toUpperCase();
}
